packages/react/src/components/Layer/index.tsx

```diff
@@ -26,33 +25,80 @@ export function useLayer() {
   };
 }
 
-export const Layer = React.forwardRef(function Layer(
+interface LayerBaseProps {
+  /**
+   * Provide child elements to be rendered inside of `Theme`
+   */
+  children?: React.ReactNode;
+
+  /**
+   * Provide a custom class name to be used on the outermost element rendered by
+   * the component
+   */
+  className?: string;
+
+  /**
+   * Specify the layer level and override any existing levels based on hierarchy
+   */
+  level?: LayerLevel;
+}
+
+export type LayerProps<T extends React.ElementType> = PolymorphicProps<
+  T,
+  LayerBaseProps
+>;
+
+export interface LayerComponent {
+  <T extends React.ElementType>(
+    props: LayerProps<T>,
+    context?: any
+  ): React.ReactElement<any, any> | null;
+}
+
+const LayerRenderFunction = React.forwardRef(function Layer<
+  T extends React.ElementType
+>(
   {
-    as: BaseComponent = 'div',
+    as,
     className: customClassName,
     children,
     level: overrideLevel,
     ...rest
-  },
-  ref
+  }: LayerProps<T>,
+  ref: React.Ref<unknown>
 ) {
   const contextLevel = React.useContext(LayerContext);
   const level = overrideLevel ?? contextLevel;
   const prefix = usePrefix();
   const className = cx(`${prefix}--layer-${levels[level]}`, customClassName);
-  // The level should be between 0 and MAX_LEVEL
-  const value = Math.max(0, Math.min(level + 1, MAX_LEVEL));
+  // The level should be between MIN_LEVEL and MAX_LEVEL
+  const value = Math.max(
+    MIN_LEVEL,
+    Math.min(level + 1, MAX_LEVEL)
+  ) as LayerLevel;
+
+  let component: React.ElementType = 'div';
+  if (as) {
+    component = as;
+  }
 
   return (
     <LayerContext.Provider value={value}>
-      <BaseComponent ref={ref} {...rest} className={className}>
-        {children}
-      </BaseComponent>
+      {React.createElement(
+        component,
+        {
+          ref,
+          ...rest,
+          className,
+        },
+        children
+      )}
```

Comment:
Instead of doing this, could we instead define `as?: keyof ReactHTML;` within `LayerBaseProps`? There's quite a few usages of the `BaseComponent` pattern within .tsx if you do a "find all" in the codebase. I'd just like to keep this consistent if possible.

Reply:
Hey, thanks for the comment! I've changed React.createElement to BaseComponent approach. Although I left PolymorphicProps for `as` prop. According to PropTypes it used to be a function or string or elementType. `keyof ReactHTML` will narrow it down to HTML tags only. Please let me know if this is the behavior you would prefer.